**Primitive Data Types**

Primitive data types are representations of numbers, alpha-numeric characters, and logic that of are part of the definition of the Java language. Primitive data types in most languages are closely related to how these
types are represented by the hardware (memory cards) in your computer. Most of the primitive data types exist
because memory and computation were very expensive 20 or 30 years ago, so using smaller representations
reduced memory use and computation. For FRC robotics programming there is no reason to favor these legacy data types, and I have highlighted the data types that we most commonly use in robotics:

* `byte` - a small memory footprint (1 byte) representation of integers
  from -128 to 127 - we generally don't worry about memory size for this in robotics, and use **`long`**.
* `short` - a small memory footprint (2 byte) representation of integers
  from -32,768 to 32,767 - we
  generally don't worry about memory size for this in robotics, and use **`long`**.
* `int` - a smaller memory footprint (4 byte) representation of integers
  from -2,147,483,648 to 2,147,483,647 - we
  generally don't worry about memory size for this in robotics, and use **`long`**.
* **`long`** (8 byte) represents -9,223,372,036,854,775,808 to 9,223,372,036,854,775,807, has alomost no additional 
  memory cost, so we normally use this for any integer (whole number) values.
* `float` - a smaller memory footprint (4 byte) representation of a real number from 1.4e-45 to 3.4e+38 - we
  generally don't worry about memory size for this in robotics, and use **`double`**.
* **`double`** - (8 byte) represents 4.9e-324 to 1.8e+308 with a precision of 15 digits and we mostly use this in robotics any time we need a real number.
* **`boolean`** - a `true` or `false` value.
* `char` - a single alpha-numeric character, which we seldom directly use in robotics. Generally we use
  **`String`** when we want a text string. We will talk about that a little later when we talk about what
  a **`class`** is.

**Variables**

Java uses named variables to store (remember) data values. The names must start with an alphabetic
character, can be any length, and cannot be Java language keywords. By convention variables start
with a lower case letter and are camel case. Variables that are constants are upper case with `'_'`
between words. For example a forward speed variable might be named `forwardSpeed` and field position
might be `fieldPositionX` and `fieldPositionY`. Constants are something that does not change like
`ROBOT_WIDTH` or `ROBOT_LENGTH`. 

The declaration of a varable is of the form:

\<decorators\> \<type\> \<variable name\>;

or

\<decorators\> \<type\> \<variable name\> = \<value\>;

The \<decorators\> are optional, these are some simple declarations of variables:

In [5]:
// set some variables
long encoderTics = 57;
double forwardSpeed = 0.8;
boolean shooterEnabled = false;

// print the values
System.out.println("encoderTics = " + encoderTics);
System.out.println("forwardSpeed = " + forwardSpeed);
System.out.println("shooterEnabled = " + shooterEnabled);

encoderTics = 57
forwardSpeed = 0.8
shooterEnabled = false


Once the variables are declared, their values can be changed:

In [6]:
encoderTics = -15;
forwardSpeed = 1.0;
shooterEnabled = true;

// print the values
System.out.println("encoderTics = " + encoderTics);
System.out.println("forwardSpeed = " + forwardSpeed);
System.out.println("shooterEnabled = " + shooterEnabled);

encoderTics = -15
forwardSpeed = 1.0
shooterEnabled = true


So what happens if you try to set a variable before you declare it?


In [9]:
testVariable = 567;

CompilationException: 

What happens if you try to set a variable to a value of the wrong data type?

In [10]:
encoderTics = true;

CompilationException: 

In [11]:
encoderTics = 27.558;

CompilationException: 

Constants are use when you have some value like robot length, robot width, Id of the motor controller on
the shooter upper roller, etc. that you will use in multiple places; will be constant the whole time code is running on the robot; but you really don't want to type in the actual value because it may change during the season as
things are changed on the robot. If you typed in actual value wherever it was used then making a change would require finding everyplace that number was used, verifying that where that number was used it actually does, for example, refer to the motor controller on the shooter upper roller, and then changing all of those. Changing a constant
is far less error prone.

When we create constants we use the decorator **`final`** to indicate that the value set is the final value and cannot be changed. For example:

In [12]:
// set some constant variables
final double DRIVE_LENGTH = 0.590;
final double DRIVE_WIDTH = 0.590;
final long SHOOTER_UPPER_ROLLER_MOTOR = 7;

// print the values
System.out.println("DRIVE_LENGTH = " + DRIVE_LENGTH);
System.out.println("DRIVE_WIDTH = " + DRIVE_WIDTH);
System.out.println("SHOOTER_UPPER_ROLLER_MOTOR = " + SHOOTER_UPPER_ROLLER_MOTOR);

DRIVE_LENGTH = 0.59
DRIVE_WIDTH = 0.59
SHOOTER_UPPER_ROLLER_MOTOR = 7


What happens if you try to change a constant?

In [19]:
DRIVE_LENGTH = 0.7;
System.out.println("DRIVE_LENGTH = " + DRIVE_LENGTH);

DRIVE_LENGTH = 0.7


Well, that's interesting the **`final`** decorator is not observed in this interpreted environment. If
I do this in real code in an IDE, the java compiler gives me the error:

```
java: cannot assign a value to final variable DRIVE_LENGTH
```

In [20]:
final double MAX_SPEED;
System.out.println("MAX_SPEED = " + MAX_SPEED);

MAX_SPEED = 0.0


Again,the **`final`** decorator is not observed in this interpreted environment. If
I do this in real code in an IDE, the java compiler gives me the error:

```
java: variable MAX_SPEED not initialized in the default constructor
```

We will revisit this when we start working with the concept of a **`class`**. It may be the case that we just don't have enough context in these examples.

**Operators and Math**

Data types have operators to help you do things like arithmetic, comparison, and logic.

Numbers have the traditional **`+`**, **`-`**, **`*`**, **`/`** operators
for addition, subtraction, multiplication, and division. With numbers, there is
also a **`Math`** library of common functions like absolute value, trigonometry functions (very useful with
the robot inertial navigation board). There are other operators (see
[Summary of Operators](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/opsummary.html)),
but my view on these is similar
to a lot of data types - they are legacy operators primarily related to memory size and computation, and
are generally not useful for FRC programming.

So let's look at some examples

In [42]:
double two = 2.0;
double four = 4.0;
double five = 5.0;
double twelve = 12.0;

System.out.println("--------------------");
System.out.println("2.0 + 4.0 = " + (2.0 + 4.0));
System.out.println("two + 4.0 = " + (two + 4.0));
System.out.println("two + four = " + (two + four));
System.out.println("--------------------");
System.out.println("5.0 - 4.0 = " + (5.0 - 4.0));
System.out.println("five - 4.0 = " + (five - 4.0));
System.out.println("five - four = " + (five - four));
System.out.println("--------------------");


--------------------
2.0 + 4.0 = 6.0
two + 4.0 = 6.0
two + four = 6.0
--------------------
5.0 - 4.0 = 1.0
five - 4.0 = 1.0
five - four = 1.0
--------------------


There is a particular syntax with math that I found really confusing when I first started programming, and that
was the case where I was tracking something like `fieldPositionX` and `fieldPositionY` and in the last command
step the robot moved `dX`, and `dY`. And I wanted to update the field position by adding the `dX` and `dY` for
this step. So my first solution to this was:

In [44]:
double fieldPositionX = 0.0;
double fieldPositionY = 0.0;

// look at robot direction and speed and compute the position change
double dX = 0.1;
double dY = 0.3;

// compute the field new positions
double newFieldPositionX = fieldPositionX + dX;
double newFieldPositionY = fieldPositionY + dY;

// set the saved field position to the new position
fieldPositionX = newFieldPositionX;
fieldPositionY = newFieldPositionY;
System.out.println("field position = (" + fieldPositionX + "," + fieldPositionY + ")");

field position = (0.1,0.3)


But, there is a condensed form of the last two steps:

In [45]:
// look at robot direction and speed and compute the position change
double dX = 0.2;
double dY = 0.2;

// update the saved field position
fieldPositionX = fieldPositionX + dX;
fieldPositionY = fieldPositionY + dY;
System.out.println("field position = (" + fieldPositionX + "," + fieldPositionY + ")");

field position = (0.30000000000000004,0.5)


Two things here:

* What does it mean to do `fieldPositionX = fieldPositionX + dX;`? Why is `fieldPositionX` on both sides of the
  equation? This can't be valid in any algibra I learned! So, yeah, this is pretty wierd. The right side of the
  `=` is evaluated first - it says get the value of `fieldPositionX` from memory (this was 0.1 after the
  previous cell), and add `dX` which is 0.2 giving us a result of 0.3, and save it as `fieldPositionX`
* The `fieldPositionX` is 0.30000000000000004 not the expected 0.3; what gives? Well, earlier we mentioned that
  a **`double`** has about 15 digits of precision. The deal is that we are taking the number (0.2) in the code and
  turning it into 64 1's and 0's for the computer - and there will be a little bit of round-off error every time
  we do that. This is one reason we like **`double`** for robotics - that there is way less round-off error
  than if we use **`float`**. A typical competition is 150 seconds, (2 min, 30 sec) and we update every 20ms
  (50 times/second) - so we have 6000 accumulations of error during a competition. Using **`double`** for
  every computation reduces the buildup of round-off errors in the 6000 cycles of a competition.
  
The next step in making the code more compact is to use the **`+=`** operator

In [46]:
// look at robot direction and speed and compute the position change
double dX = 0.25;
double dY = 0.5;

// update the saved field position
fieldPositionX += dX;
fieldPositionY += dY;
System.out.println("field position = (" + fieldPositionX + "," + fieldPositionY + ")");

field position = (0.55,1.0)


OK, what just happened there? The **`+=`** assignment says get the value of `fieldPositionX`, add `dX`, and
store the result back into `fieldPositionX`. Once you have the secret decoder ring, these compact
representations are pretty easy to read amd make a lot of sense.

Let's talk a little bit about the **`Math`** library. This library is primarily for **`double`**
values. It includes both useful math constants and functions like sine, cosine, tangent, arc-sine,
arc-cosine, arc-tangent radian-degree conversions, power, root, and log functions, etc. Let's start
by looking at the **`Math.PI`** constant. Note that it follows the convention that the constant name,
**`PI`**, is capitalized. NOTE ALSO, if we try to change the value of **`Math.PI`** we get the expected
error for editing constants - so, I think the examples in the previous section did not have enough
context for the **`final`** decorator to be meaningful.

In [37]:
System.out.println("Math.PI = " + Math.PI);
Math.PI = 4.0;

Math.PI = 3.141592653589793


CompilationException: 

Let's look at a robot control example where we gat a heading from the NavX, and have a heading we expect
the robot to be tracking. It turns out the NavX reports everything in degrees, but we have agreed as a team
to use SI notation and express angles in degrees. So suppose we want to compute the heading error (difference
between the expected heading and the actual heading) so we can try to correct that in our next command to
the drive:

In [52]:
double expectedHeading = 0.5238;
double navxHeading = 45.0;

double headingError = expectedHeading - Math.toRadians(navxHeading);
System.out.println("headingError (radians) = " + headingError);
System.out.println("headingError (degrees) = " + Math.toDegrees(headingError));

headingError (radians) = -0.26159816339744824
headingError (degrees) = -14.988470691047477


Let's look at another robot control example. The control stick on the gamepad returns 0.0 when centered; -1.0 when full left; and 1.0 when full right. When the stick is halfway towards full right it returns about 0.5 which is what
we expect. However, if a driver needs fine control in the heat of competition to line up the robot to score, the driver has trouble moving just a little bit. So the driver wants better control for fine movements.

One way to do that is a power function. For example, if I square the input value fronm the stick, 0.0 is still 0.0;
1.0 is still 1.0; but the halfway value of 0.5 is now 0.25 - giving better control for small movements. Cool, but if I'm going left -0.5 squared is 0.25, so the robot is going the wrong direction. How do I fix that? or even allow a fractional power like 1.5 or 2.5.

Again, the **`Math`** library helps us out:

In [55]:
double stickX = -0.5;
double signMultiplier = Math.signum(stickX);
double absStickX = Math.abs(stickX);
double useX = Math.pow(absStickX, 2.0) * signMultiplier;

System.out.println("useX = " + useX);

useX = -0.25


One more thing about math - order of operations and parenthesis. It is often the case that you get
a comple math expression like

In [56]:
3.0 * 12.0 / 56.0 - 8.0 * 12.0

-95.35714285714286

Which is really confusing there is an order of operations that explains what to expect, but, it
is non-obvious, requires you remember the order of obperations and then apply them to figure
out what is going on. Use parenthesis so it is clear what the order of oberations isintended to be:

In [57]:
(3.0 * 12.0) / (56.0 - (8.0 * 12.0))

-0.9